xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceTradeService.java

```diff
@@ -206,23 +205,42 @@ public boolean cancelOrder(String orderId) {
   public boolean cancelOrder(CancelOrderParams params) throws IOException {
     try {
       if (!(params instanceof CancelOrderByCurrencyPair)
-          && !(params instanceof CancelOrderByIdParams)) {
+          || !(params instanceof CancelOrderByIdParams)) {
         throw new ExchangeException(
             "You need to provide the currency pair and the order id to cancel an order.");
       }
       CancelOrderByCurrencyPair paramCurrencyPair = (CancelOrderByCurrencyPair) params;
       CancelOrderByIdParams paramId = (CancelOrderByIdParams) params;
-      super.cancelOrder(
-          paramCurrencyPair.getCurrencyPair(),
-          BinanceAdapters.id(paramId.getOrderId()),
-          null,
-          null);
+      String orderId = paramId.getOrderId();
+      MarginAccountType marginAccountType =
+              params instanceof BinanceCancelOrderParams ? ((BinanceCancelOrderParams) params).getMarginAccountType()
+                      : includeMarginAccountTypeInOrderId ? getMarginAccountTypeFromOrderId(orderId)
+                      : null;
+      if (marginAccountType == null) {
+        super.cancelOrder(
+                paramCurrencyPair.getCurrencyPair(),
+                BinanceAdapters.id(orderId),
+                null,
+                null);
+      } else {
+        super.cancelMarginOrder(
+                paramCurrencyPair.getCurrencyPair(),
+                marginAccountType,
+                BinanceAdapters.id(orderId),
+                null,
+                null);
+      }
       return true;
     } catch (BinanceException e) {
       throw BinanceErrorAdapter.adapt(e);
     }
   }
 
+  private MarginAccountType getMarginAccountTypeFromOrderId(String orderId) {
```

Comment:
using order id to smuggle the margin accounts id look hacky

is this some kind of convention that comes from binance?

could you point to the documentation where its described?

Reply:
This is not a part of official documentation. It's a hack added by me to allow using regular cancelOrder/getOrder API for margin orders without any special parameter types. And it's optional though